processor/src/main/java/com/linecorp/decaton/processor/runtime/ProcessorSubscription.java

```diff
@@ -103,15 +113,63 @@ private void updateState(SubscriptionStateListener.State newState) {
+                // another async commit without waiting the first one to complete.
+                // This would be harmless if `decaton.commit.interval.ms` is set to sufficiently large, but otherwise
+                // we would make abusive offset commits despite there's no progress in processing.
+                log.debug("Skipping commit due to another async commit is currently in-flight");
+                return;
+            }
+            consumer.commitAsync(commitOffsets, (offsets, exception) -> {
+                asyncCommitInFlight = false;
+                if (exception != null) {
+                    log.warn("Offset commit failed asynchronously", exception);
+                    return;
+                }
+                // Exception raised from the commitAsync callback bubbles up to Consumer.poll() so it kills this
+                // subscription thread.
+                // Basically the exception thrown from here is considered a bug, but failing to commit offset itself
+                // isn't usually critical for consumer's continuation availability so we grab it here widely.
+                try {
+                    // Some thread and timing safety consideration about below operation.
+                    // Thread safety:
+                    // Below operation touches some thread-unsafe resources such as ProcessingContexts map and
+                    // variable storing committed offset in ProcessingContext but we can assume this callback is
+                    // thread safe because in all cases a callback for commitAsync is called from inside of
+                    // Consumer.poll() which is called only by this (subscription) thread.
+                    // Timing safety:
+                    // At the time this callback is triggered there might be a change in ProcessingContexts map
+                    // underlying contexts. It would occur in two cases:
+                    // 1. When group rebalance is triggered and the context is dropped by revoke at partitionsRevoked().
+                    // 2. When dynamic processor reload is triggered and the context is renewed by
+                    // PartitionContexts.maybeHandlePropertyReload.
+                    // The case 2 is safe (safe to keep updating committed offset in renewed PartitionContext) because
+                    // it caries previously consuming offset without reset.
+                    // The case 1 is somewhat suspicious but should still be safe, because whenever partition revoke
+                    // happens it calls commitSync() through onPartitionsRevoked(). According to the document of
+                    // commitAsync(), it is guaranteed that its callback is called before subsequent commitSync()
+                    // returns. So when a context is dropped from PartitionContexts, there should be no in-flight
+                    // commitAsync().
+                    contexts.updateCommittedOffsets(commitOffsets);
```

Comment:
since this now async, different threads can call this. I suppose the actual `PartitionContext.lastCommittedOffset` should now be volatile?

Reply:
No it can't. Please read through this comment explaining what I observed from reading kafka code: https://github.com/line/decaton/pull/53/files#diff-bd5dd2f1469e6022198c76787fd059eaR147

Reply:
Believe it or not I did not notice your novel over here. Sorry about that..
In any case, I would feel safer with an actual test that checks the thread in the callback; although it sounds unlikely that Kafka client would change its threading model?

Reply:
It is a case of acute TLDRitis, I'm afraid 😭

Reply:
>  although it sounds unlikely that Kafka client would change its threading model?

KafkaConsumer itself is basically "for a single thread" component. It doesn't have background IO thread (except a thread dedicated for heartbeat), and from what we can see in implementation of commitAsync callback it is obviously controlled it's triggering timing and location with intention:
* https://github.com/apache/kafka/blob/6f3ceea5c61d90d4075ed5c5a22bde33142f7ec7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L1452 (callback triggered here)
* https://github.com/apache/kafka/blob/6f3ceea5c61d90d4075ed5c5a22bde33142f7ec7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L900 (the entrypoint calling .invoke())
* `invokeCompletedOffsetCommitCallbacks` is called only by `poll`, `commitXX` and `close`

Reading this implementation I think it will be unlikely that they change implementation in the future without major, noticable breaking change.

> In any case, I would feel safer with an actual test that checks the thread in the callback;

Well it's not that easy to add a test that confirms it but I just came up with one cheap assert that we can log when it happens at least. f748ead

Reply:
Thanks. 👍